tracer/src/Datadog.Trace/Ci/TestModule.cs

```diff
@@ -0,0 +1,315 @@
+    /// <param name="name">Test module name</param>
+    /// <param name="framework">Testing framework name</param>
+    /// <param name="frameworkVersion">Testing framework version</param>
+    /// <param name="startDate">Test session start date</param>
+    /// <returns>New test session instance</returns>
+    public static TestModule Create(string name, string? framework, string? frameworkVersion, DateTimeOffset? startDate)
+    {
+        return new TestModule(name, framework, frameworkVersion, startDate);
+    }
+
+    /// <summary>
+    /// Sets a string tag into the test
+    /// </summary>
+    /// <param name="key">Key of the tag</param>
+    /// <param name="value">Value of the tag</param>
+    public void SetTag(string key, string? value)
+    {
+        _span.SetTag(key, value);
+    }
+
+    /// <summary>
+    /// Sets a number tag into the test
+    /// </summary>
+    /// <param name="key">Key of the tag</param>
+    /// <param name="value">Value of the tag</param>
+    public void SetTag(string key, double? value)
+    {
+        _span.SetMetric(key, value);
+    }
+
+    /// <summary>
+    /// Set Error Info
+    /// </summary>
+    /// <param name="type">Error type</param>
+    /// <param name="message">Error message</param>
+    /// <param name="callStack">Error callstack</param>
+    public void SetErrorInfo(string type, string message, string? callStack)
+    {
+        var span = _span;
+        span.Error = true;
+        span.SetTag(Trace.Tags.ErrorType, type);
+        span.SetTag(Trace.Tags.ErrorMsg, message);
+        if (callStack is not null)
+        {
+            span.SetTag(Trace.Tags.ErrorStack, callStack);
+        }
+    }
+
+    /// <summary>
+    /// Set Error Info from Exception
+    /// </summary>
+    /// <param name="exception">Exception instance</param>
+    public void SetErrorInfo(Exception exception)
+    {
+        _span.SetException(exception);
+    }
+
+    /// <summary>
+    /// Close test module
+    /// </summary>
+    public void Close()
+    {
+        Close(null);
+    }
+
+    /// <summary>
+    /// Close test module
+    /// </summary>
+    /// <param name="duration">Duration of the test module</param>
+    public void Close(TimeSpan? duration)
+    {
+        if (Interlocked.Exchange(ref _finished, 1) == 1)
+        {
+            return;
+        }
+
+        var span = _span;
+
+        // Calculate duration beforehand
+        duration ??= span.Context.TraceContext.ElapsedSince(span.StartTime);
+
+        var remainingSuites = Array.Empty<TestSuite>();
+        lock (_suites)
+        {
+            if (_suites.Count > 0)
+            {
+                remainingSuites = _suites.Values.ToArray();
+            }
+        }
+
+        foreach (var suite in remainingSuites)
+        {
+            suite.Close();
+        }
+
+        // Update status
+        Tags.Status ??= TestTags.StatusPass;
+
+        // Finish if agentless is enabled
+        if (CIVisibility.Settings.Agentless)
```

Comment:
Will this span get finished when we don't have agentless enabled?

Reply:
nope, APM intake doesn't support test suite visibility. But also I didn't like this way to avoid sending those events in case of the Agent based solution, and I changed in https://github.com/DataDog/dd-trace-dotnet/pull/3276/commits/1804c381ada9ab0c3c2fcbab2e47173544f57df9 with a new `TestSuiteVisibilityProcessor`

Reply:
I like it!